uniffi_bindgen/src/bindings/gecko/gen_gecko.rs

```diff
@@ -0,0 +1,355 @@
+}
+
+/// An implementation file generated for a namespace with top-level functions.
+#[derive(Template)]
+#[template(syntax = "cpp", escape = "none", path = "InterfaceTemplate.cpp")]
+pub struct Interface<'config, 'ci, 'obj> {
+    config: &'config Config,
+    ci: &'ci ComponentInterface,
+    obj: &'obj Object,
+}
+
+impl<'config, 'ci, 'obj> Interface<'config, 'ci, 'obj> {
+    pub fn new(config: &'config Config, ci: &'ci ComponentInterface, obj: &'obj Object) -> Self {
+        Self {
+            config: config,
+            ci,
+            obj,
+        }
+    }
+}
+
+/// Filters for our Askama templates above. These output C++, XPIDL, and
+/// WebIDL.
+mod filters {
+    use super::*;
+    use std::fmt;
+
+    /// Declares a WebIDL type in the interface for this library.
+    pub fn type_webidl(type_: &Type) -> Result<String, askama::Error> {
+        Ok(match type_ {
+            Type::Int8 => "byte".into(),
+            Type::UInt8 => "octet".into(),
+            Type::Int16 => "short".into(),
+            Type::UInt16 => "unsigned short".into(),
+            Type::Int32 => "long".into(),
+            Type::UInt32 => "unsigned long".into(),
+            Type::Int64 => "long long".into(),
+            Type::UInt64 => "unsigned long long".into(),
+            Type::Float32 => "float".into(),
+            // Note: Not `unrestricted double`; we don't want to allow NaNs
+            // and infinity.
+            Type::Float64 => "double".into(),
+            Type::Boolean => "boolean".into(),
+            Type::String => "DOMString".into(),
+            Type::Enum(name) | Type::Record(name) | Type::Object(name) => class_name_webidl(name)?,
+            Type::Error(name) => panic!("[TODO: type_webidl({:?})]", type_),
+            Type::Optional(type_) => format!("{}?", type_webidl(type_)?),
+            Type::Sequence(type_) => format!("sequence<{}>", type_webidl(type_)?),
+            Type::Map(type_) => format!("record<DOMString, {}>", type_webidl(type_)?),
+        })
+    }
+
+    /// Declares a C type in the `extern` declarations.
+    pub fn type_ffi(type_: &FFIType) -> Result<String, askama::Error> {
+        Ok(match type_ {
+            FFIType::Int8 => "int8_t".into(),
+            FFIType::UInt8 => "uint8_t".into(),
+            FFIType::Int16 => "int16_t".into(),
+            FFIType::UInt16 => "uint16_t".into(),
+            FFIType::Int32 => "int32_t".into(),
+            FFIType::UInt32 => "uint32_t".into(),
+            FFIType::Int64 => "int64_t".into(),
+            FFIType::UInt64 => "uint64_t".into(),
+            FFIType::Float32 => "float".into(),
+            FFIType::Float64 => "double".into(),
+            FFIType::RustBuffer => "RustBuffer".into(),
+            FFIType::RustString => "char*".into(),
+            FFIType::RustError => "NativeRustError".into(),
+            FFIType::ForeignStringRef => "const char*".into(),
+        })
+    }
+
+    pub fn ret_type_ffi(type_: &Type) -> Result<String, askama::Error> {
```

Comment:
Needs a comment—this is used in `macros.cpp`, to generate the specialization declaration for lifting return values from the FFI call, and for declaring the return type of an FFI call.